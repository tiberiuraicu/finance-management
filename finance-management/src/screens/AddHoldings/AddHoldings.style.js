import { StyleSheet } from "react-native";
export default StyleSheet.create({
  inputContainer: {
   
    paddingTop: "15%",
    marginTop: "20%",
    marginLeft: "5%",
    marginRight: "5%",
    backgroundColor: "#F2FCFE",
    borderRadius: 10,
  },
  searchTermInput:{
    borderRadius: 10,
    height: "14%",
    borderWidth: 1,
    margin: 11,
    paddingLeft: 10,
  },
  
});

export let searchTermInputText =(symbol)=>{
  return {
    fontSize: 18,
    marginTop: 20,
    color: symbol.value === "" ? "#788793" : "black",
  }
}
