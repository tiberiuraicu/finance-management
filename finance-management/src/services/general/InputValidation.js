import { inputModel } from "../../util/models";

// @info : validates a given value depending 
// on the given optional parameters 
// (it can test a value using one or more checks) 
export const inputValidator = (
  valueToBeTested,
  setInputProperties,
  { notEmpty = false, isNumber = false } = {}
) => {
  var inputProperties = inputModel()

  // only tests if the valueToBeTested does not have to be empty
  if (notEmpty)
    inputProperties = notEmptyCheck(valueToBeTested, inputProperties);

  // only tests if the valueToBeTested should be a number
  if (isNumber)
    inputProperties = isNumberCheck(valueToBeTested, inputProperties);

  // if none of the above conditions are met => no error message will be shown
  if (inputProperties.errorMessage === "")
    inputProperties = {
      ...inputProperties,
      validation: true,
    };

  // sets the value for a specific field(ex : price, numberOfShares)
  // and also the other properties that were changed above
  setInputProperties({ ...inputProperties, value: valueToBeTested });
};

// checks if the field is empty => error
const notEmptyCheck = (valueToBeTested, inputProperties) => {
  if (valueToBeTested === "")
    inputProperties = {
      ...inputProperties,
      validation: false,
      errorMessage: "This field cannot be empty",
    };

  return inputProperties;
};

// checks if the value typed is a number => error
const isNumberCheck = (valueToBeTested, inputProperties) => {
  
  // regex which allows just numbers and "."
  const re = /^[0-9]*\.?[0-9]*$/;

  if (!re.test(valueToBeTested))
    inputProperties = {
      ...inputProperties,
      validation: false,
      errorMessage: "Only numbers accepted",
    };

  return inputProperties;
};
