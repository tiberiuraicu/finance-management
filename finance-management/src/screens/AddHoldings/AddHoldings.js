import React, { useState, useEffect } from "react";
import {
  View,
  useWindowDimensions,
  TouchableOpacity,
  Text,
  Alert,
} from "react-native";

// importing custom components
import TextInputCustom from "../../components/TextInputCustom";
import WrapperScrollView from "../../components/WrapperScrollView";
import CustomButton from "../../components/CustomButton";
import ErrorMessage from "../../components/ErrorMessage";
import SymbolSearchModal from "../../components/SymbolSearchModal";

// importing styles
import styles, { searchTermInputText } from "./AddHoldings.style";

// importing models
import { inputModel, transactionModel } from "../../util/models";

//importing services
import {
  addHoldings,
  editHoldings,
} from "../../services/screenServices/AddHoldings.services";
import { inputValidator } from "../../services/general/InputValidation";

import { useIsFocused } from "@react-navigation/native";

const AddHoldings = ({ route, navigation }) => {
  const isFocused = useIsFocused();

  const { isAddHolding, symbolForEdit } = route.params;
  // get the window height
  const windowHeight = useWindowDimensions().height;

  // hooks for the input fields
  const [symbol, setSymbol] = useState(inputModel());
  const [price, setPrice] = useState(inputModel());
  const [numberOfShares, setNumberOfShares] = useState(inputModel());

  // hook for the modal visibility
  const [searchModalVisibility, setSearchModalVisibility] = useState(false);

  // calls a service function which verifies if the
  // value is correct and then sets the price value
  const setPriceHandler = (valueToBeTested) => {
    inputValidator(valueToBeTested, setPrice, {
      notEmpty: true,
      isNumber: true,
    });
  };

  // calls a service function which verifies if the
  // value is correct and then sets the nunber of shares value
  const setNumberOfSharesHandler = (valueToBeTested) => {
    inputValidator(valueToBeTested, setNumberOfShares, {
      notEmpty: true,
      isNumber: true,
    });
  };

  // calls a service function which adds a new transaction
  // to portfolio. It will execute only if the value from
  // each field is correct
  const addHoldingsHandler = async () => {
    // if the values from every field are correct
    if (symbol.validation && price.validation && numberOfShares.validation) {
      // creates a new transaction object
      const transaction = transactionModel(price.value, numberOfShares.value);

      //calls the addHoldings method
      let result = await addHoldings(transaction, symbol.value);

      if (result) Alert.alert("", "Transaction added");
      else Alert.alert("", "Transaction failed");
    }
  };

  const editHoldingsHandler = async () => {
    // if the values from every field are correct
    if (symbol.validation && price.validation && numberOfShares.validation) {
      // creates a new transaction object
      const transaction = transactionModel(price.value, numberOfShares.value);

      //calls the editHoldings method
      let result = await editHoldings(transaction, symbol.value);

      if (result) Alert.alert("", "Holding updated");
      else Alert.alert("", "Update failed");
    }
  };
  useEffect(() => {
    if (!isAddHolding) {
      setSymbol({ value: symbolForEdit, validation: true });
    }
  }, [isFocused]);

  return (
    <WrapperScrollView>
      {/* the view which contains the inputs */}
      <View
        style={{
          ...styles.inputContainer,

          // give it a specific height based on screen height
          minHeight: Math.round(windowHeight / 1.5),
        }}
      >
        <Text style={styles.title}>
          {isAddHolding ? "New transaction" : "Edit holding"}
        </Text>
        {/* looks like an input
         => when it is pressed a modal appears
            which lets the user search for a share */}
        <TouchableOpacity
          onPress={() => {
            if (isAddHolding) setSearchModalVisibility(true);
          }}
          style={styles.searchTermInput}
        >
          {/* will have either a placeholder value or 
          the selected share symbol(from the modal) */}
          <Text style={searchTermInputText(symbol)}>
            {symbol.value !== "" ? symbol.value : "Search term"}
          </Text>
        </TouchableOpacity>

        {/* the modal in which the user searches for a specific share */}
        {searchModalVisibility && (
          <SymbolSearchModal
            searchModalVisibility={searchModalVisibility}
            setSearchModalVisibility={setSearchModalVisibility}
            symbol={symbol}
            setSymbol={setSymbol}
          ></SymbolSearchModal>
        )}

        {/* symbol error message */}
        <ErrorMessage errorTruthValue={symbol.validation}>
          {symbol.errorMessage}
        </ErrorMessage>

        {/* the user can enter the price 
        for which an asset has been bought */}
        <TextInputCustom
          placeholder="Price"
          value={price.value}
          onChangeText={setPriceHandler}
          keyboardType="numeric"
        ></TextInputCustom>

        {/* price error message */}
        <ErrorMessage errorTruthValue={price.validation}>
          {price.errorMessage}
        </ErrorMessage>

        {/* the user can enter the 
       number of shares bought */}
        <TextInputCustom
          placeholder="Number of shares"
          value={numberOfShares.value}
          onChangeText={setNumberOfSharesHandler}
          keyboardType="numeric"
        ></TextInputCustom>

        {/* number of shares error message */}
        <ErrorMessage errorTruthValue={numberOfShares.validation}>
          {numberOfShares.errorMessage}
        </ErrorMessage>

        {isAddHolding ? (
          <CustomButton
            text="ADD HOLDING"
            onPress={addHoldingsHandler}
          ></CustomButton>
        ) : (
          <CustomButton
            text="EDIT HOLDING"
            onPress={editHoldingsHandler}
          ></CustomButton>
        )}
      </View>
    </WrapperScrollView>
  );
};

export default AddHoldings;
