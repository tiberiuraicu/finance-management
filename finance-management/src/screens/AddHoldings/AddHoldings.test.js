jest.mock("../../services/general/InputValidation", () => ({
  inputValidator: jest.fn(),
}));
jest.mock("../../services/screenServices/AddHoldings.services", () => ({
  addHoldings: jest.fn().mockReturnValueOnce(false).mockReturnValueOnce(true),
}));

jest.mock("react", () => {
  const originalModule = jest.requireActual("react");

  return {
    ...originalModule,
    useState: jest
      .fn()
      .mockImplementation(() => [
        { value: "", validation: true, errorMessage: "" },
        () => {},
      ]),
  };
});

import AddHoldings from "./AddHoldings.js";
import { shallow, mount } from "enzyme";
import { inputValidator } from "../../services/general/InputValidation";
import { addHoldings } from "../../services/screenServices/AddHoldings.services";
import { Alert } from "react-native";

describe("AddHoldings screen tests", () => {
  it("renders the component", () => {
    const wrapper = shallow(<AddHoldings />);
    expect(wrapper).toMatchSnapshot();
  });
  it("calls the input validator", () => {
    const wrapper = shallow(<AddHoldings />);
    wrapper.find("TextInputCustom").at(0).simulate("changeText", "1");
    expect(inputValidator).toBeCalled();
    wrapper.find("TextInputCustom").at(1).simulate("changeText", "1");
    expect(inputValidator).toBeCalled();
  });

  it("calls the addHoldings method on button press (with transaction failed alert)", async () => {
    jest.spyOn(Alert, "alert");

    const wrapper = mount(<AddHoldings />);
    await wrapper.find("CustomButton").at(0).props().onPress();

    expect(addHoldings).toBeCalled();

    // it gives 'Transzction failed' because there
    // is no transaction object and an error is thrown
    // => it gets to this point because the validations
    // for inputs have been mocked to be true
    expect(Alert.alert).toHaveBeenCalledWith("", "Transaction failed");
  });

  it("calls the addHoldings method on button press (with transaction added alert)", async () => {
    jest.spyOn(Alert, "alert");

    const wrapper = mount(<AddHoldings />);
    await wrapper.find("CustomButton").at(0).props().onPress();

    expect(addHoldings).toBeCalled();

    // it gives 'Transzction failed' because there
    // is no transaction object and an error is thrown
    // => it gets to this point because the validations
    // for inputs have been mocked to be true
    expect(Alert.alert).toHaveBeenCalledWith("", "Transaction added");
  });
});
