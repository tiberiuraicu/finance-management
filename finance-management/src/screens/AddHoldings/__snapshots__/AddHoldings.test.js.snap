// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddHoldings screen tests renders the component 1`] = `ShallowWrapper {}`;
